base/static/base/js/CGIMailForm.js

```diff
@@ -0,0 +1,411 @@
+            </div>
+          ))}
+        </div>
+      );
+    }
+    return buildField(elm, state, handleChange);
+  });
+};
+
+const SectionTitle = (props) => {
+  const { title } = props;
+  return <h2>{title}</h2> || '';
+};
+
+const Section = (props) => {
+  const {
+    title, elements, handleChange, state, hidden,
+  } = props;
+  if (hidden) {
+    return (
+      <FormElements elements={elements} handleChange={null} state={state} />
+    );
+  }
+  return (
+    <section>
+      <SectionTitle title={title} />
+      <FormElements
+        elements={elements}
+        handleChange={handleChange}
+        state={state}
+      />
+    </section>
+  );
+};
+
+Section.defaultProps = {
+  title: null,
+  hidden: null,
+};
+
+Section.propTypes = {
+  title: PropTypes.string,
+  elements: PropTypes.arrayOf(PropTypes.object).isRequired,
+  handleChange: PropTypes.func.isRequired,
+  state: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
+  hidden: PropTypes.bool,
+};
+
+const Sections = (props) => {
+  const { data, handleChange, state } = props;
+  return data.map(d => (
+    <Section
+      title={d.title}
+      elements={d.elements}
+      handleChange={handleChange}
+      state={state}
+      hidden={d.hidden || null}
+    />
+  ));
+};
+
+const LoadingItemInfo = (props) => {
+  const { loading } = props;
+  if (loading === true) {
+    return <p className="text-success">Retrieving item information...</p>;
+  }
+  return '';
+};
+
+class FormContainer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleChange = this.handleChange.bind(this);
+    this.state = {
+      submitted: false,
+      response: false,
+      itemInfo: false, // eslint-disable-line react/no-unused-state
+      isLoadingItemInfo: false,
+    };
+  }
+
+  componentDidMount() {
+    let itemServlet = ITEM_SERVLET;
+    const bib = URLPARAMS.get('bib') || null;
+    const barcode = URLPARAMS.get('barcode') || null;
+    let itemInfo = false;
+    if (bib || barcode) {
+      itemInfo = true;
+      this.setState({
+        isLoadingItemInfo: true,
+      });
+      if (bib) {
+        itemServlet = `${itemServlet}&bib=${bib}`;
+      }
+      if (barcode) {
+        itemServlet = `${itemServlet}&barcode=${barcode}`;
+      }
+    }
+    if (itemInfo) {
+      return fetch(`${itemServlet}`)
+        .then(res => res.json())
+        .then((res) => {
+          this.setState({
+            itemInfo: res, // eslint-disable-line react/no-unused-state
+            isLoadingItemInfo: false,
+          });
+        })
+        .catch((error) => {
+          console.error(error); // eslint-disable-line no-console
+        });
+    }
+    return null;
+  }
+
+  handleRedirect = (res) => {
+    if (res.status === 200) {
+      res.text().then((body) => {
+        this.setState({
+          submitted: true,
+          response: body,
+        });
+      });
+    } else {
+      // TODO - Failure message
```

Comment:
Forgotten TODO or something optional?

Reply:
I was viewing this as optional. I could never get CGIMail to return anything else. Normally when it "fails", it still gives a 200.

Reply:
Ah, interesting.